### PR TITLE
[Android] Add initial ERC721 support

### PR DESCRIPTION
**Pull request by wchen342**
Resolves https://github.com/brave/brave-browser/issues/19565.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java [hunk 1]
```diff
@@ -884,10 +885,14 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
                     TxData data =
                             Utils.getTxData("", "", "", to, Utils.toHexWei(value, 18), new byte[0]);
                     sendTransaction(data, from, "", "");
-                } else {
+                } else if (mCurrentBlockchainToken.isErc20) {
                     addUnapprovedTransactionERC20(to,
                             Utils.toHexWei(value, mCurrentBlockchainToken.decimals), from,
                             mCurrentBlockchainToken.contractAddress);
+                } else if (mCurrentBlockchainToken.isErc721) {
+                    // TODO tokenId text field
```
[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/BuySendSwapActivity.java:893]: yeah, the initial PR doesn't do Send

File: android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java [hunk 2]
```diff
@@ -143,7 +144,8 @@ public int getIcon() {
     }
 
     public String getTitle() {
-        return mTitle;
+        // ERC721 has format [Title #ID]
+        return mTitle + (isErc721() ? "" : "#" + Utils.hexToIntString(mId));
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java:148]: is that correct? as I see we do `[Title #ID]` for non ERC721 now. Shouldn't it be an opposite?

File: android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java [hunk 3]
```diff
@@ -165,4 +167,12 @@ public boolean getIsUserSelected() {
     public void setIsUserSelected(boolean isUserSelected) {
         mIsUserSelected = isUserSelected;
     }
+
+    public boolean isErc721() {
+        return mId == null || mId.trim().isEmpty();
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java:172]: that's ERC20 I think

File: android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java (hunk 2 above)
Comment (wchen342) [reply android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java:148]: The if here is actually working correctly with the conditions in `isErc721`, just the function being misnamed. But for clarity I'll flip the conditions.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/TokenUtils.java [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 package org.chromium.chrome.browser.crypto_wallet.util;
 
+import org.chromium.base.Log;
```
[the remark this answers is not in the document]
Comment (wchen342) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/TokenUtils.java]: Right, removed now.

File: android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java (hunk 3 above)
Comment (wchen342) [reply android/java/org/chromium/chrome/browser/crypto_wallet/model/WalletListItemModel.java]: Same as above, I flipped the condition.

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Pull request merged**